### ISSUE TITLE
[new feature] Edit QnA prompt for non-english texts

### ISSUE THREAD

**Issue by eostis** [weaviate/weaviate#2587]
Hi,

A client with a non-English WooCommerce site tested the OpenAI QnA module, but was not impressed by the answers.

I wondered if it could be due to the prompt written in English?
(I presumed that candidates for the question are found by a search, then sent as context to the QnA)

If so, could we:

1. Be able to monitor what has been sent to the QnA (OpenAI or transformer model) with a dedicated field in the GraphQL query ?
2. Be able to edit the prompt with a dedicated field in the GraphQL query ?

**Comment by etiennedi** [comment]
Looping in @byronvoorbach or @CShorten who both have much more experience with OpenAI prompts than me.

**Comment by CShorten** [comment]
Hey all,

1. With the new `generate` module you will be able to see the Weaviate search results passed to the OpenAI call (in this case `content`) and then the output of the OpenAI call in `result` like this:

```graphql
{
  Get{
    PodClip(
      limit: 1
      nearText:{
        concepts:["How does replication work?"]
      }
    ){
      content
      _additional{
        generate(
          task:"Rewrite this into an ELI5 summary"
          resultLanguage: German
          onSet: individualResults
        	properties :[
                  "content"
                ]
        ){
          result
        }
      }
    }
  }
}
```

2. We have discussed a template double curly bracket syntax {{ }}} such as revising the task argument to:

```graphql
task: """
Please write an ELI5 summary of the search results,

{{speaker} said {content} on {date}}

Please try to keep this summary as simple as possible.
"""
```

Could you please explain further how you would like to modify the prompt with the property @eostis? It would really help guide the thinking on this 🙏

**Comment by eostis** [comment]
I imagined that the current QnA module was building a prompt from the retrieval stage.

Like:
"question: <the question> context: <retrieved documents from the question>"

And I wondered if the template of the prompt being in English (as shown above in the example), if that could explain that a question in another language with results in the same language would been mixed to the english template, hence giving not so good results.

If so, I wondered if the template could be updated, to be translated  or adapted.

**Comment by felizalde** [comment]
Hey all, my first time here, but already playing a lot with Weaviate, great product! 
Here is the workaround I found to ask questions to content in Spanish. It may could help to what @eostis want to do.

I think the prompt that @eostis want to edit is this one: https://github.com/weaviate/weaviate/blob/7ee1d973587bf7f80566ba545d80c45ce1623ee1/modules/qna-openai/clients/qna.go#L121

```
Please answer the question according to the above context.

===
{context}
===
Q: {question}
A:
```

*The {context} value is calculated from the retrieved values of each property listed in the properties array in the ask parameter.*
*btw, the prompt is wrong it should said: `[..] according to the **below** context.`*

So once we know the prompt generated in the `qna` module, we can replicate it with the new `generative` module. 
This two queries works as the same way and will generate the same answer.

Using `qna-openai`
```gql
{
  Get {
    Article (
      limit: 1
      ask: {
        question: "Who is Stanley Kubrick?",
        properties: ["summary"]
      }
      nearText: {
        concepts: ["Who is Stanley Kubrick?"]
      }
    ) {
      summary,
      _additional {
        answer {
          hasAnswer
          result
        } 
      }
    }
  }
}
```

Using `generative-openai`

```gql
{
  Get {
    Article (
      limit: 1
      nearText: {
        concepts: ["Who is Stanley Kubrick?"]
      }
    ) {
      summary,
      _additional {
        generate(
          singleResult: {
            prompt: "Please answer the question according to the above context.\n\n===\n{summary}\n===\nQ:Who is Stanley Kubrick?\nA:"
          }
        ) {
          singleResult
        }
      }
    }
  }
}
```

If you want to change the `qna` prompt to a different language. You can just write the prompt in the new `generative` module in the language you want. For example in Spanish:

```gql
{
  Get {
    Article (
      limit: 1
      nearText: {
        concepts: ["¿Quién es Lionel Messi?"]
      }
    ) {
      summary,
      _additional {
        generate(
          singleResult: {
            prompt: "Por favor, responda la pregunta de acuerdo con el siguiente contexto.\n\n===\n{summary}\n===\nQ:¿Quién es Lionel Messi?\nA:"
          }
        ) {
          singleResult
        }
      }
    }
  }
}
```
*This to work well need the vectorizer be able to work in the specific language. OpenAI embeddings are multi-language and support a grant variety of lang. The `contextionary` module support a few ones, you can check them [here](https://weaviate.io/developers/weaviate/modules/retriever-vectorizer-modules/text2vec-contextionary#available-modules-and-languages).*

I'm not sure if the `qna-openai` module needs to allow changing the prompt template. At least not from the query. I think keeping it as simple as possible is good, for more complex prompts you can use the `generative` module.  A good place to edit it might be the [module's settings](https://weaviate.io/developers/weaviate/modules/reader-generator-modules/qna-openai#how-to-configure) in the schema definition.

Also, @CShorten would be really nice if we can include to the `_additional` results for the `openai` modules the total tokens used and the finish reason. The tokens is good to track the usage and calculate cost of each query, and the last one is a good indicative to know if the model reaches a good stopping point, even [OpenAI mention](https://platform.openai.com/docs/guides/completion/best-practices) that is a good practice to check it.

> Prefer finish_reason == "stop". When the model reaches a natural stopping point or a user provided stop sequence, it will set finish_reason as "stop". This indicates that the model has managed to connect to the suffix well and is a good signal for the quality of a completion.

Let me know if you'd like me to create a new issue about the addition of `total_tokens` and `finish_reason` to the response, as it's a bit off topic for this thread.

**Comment by eostis** [comment]
Thanks @felizalde.

I tested manually with the official prompt template you mentioned. Results are good, even when the template is english, and  context and questions different languages.

So, I guess my issues came from the client's data itself (context).

Which brings back another point: being able to debug the prompt generated during the query (#2518)
